californium-core/src/main/java/org/eclipse/californium/core/network/BaseMatcher.java

```diff
@@ -83,9 +108,136 @@ public synchronized void stop() {
 	/**
 	 * This method does nothing.
 	 * <p>
-	 * Subclasses should override this method in order to clear any internal state.
+	 * Subclasses should override this method in order to clear any internal
+	 * state.
 	 */
 	@Override
 	public void clear() {
 	}
+
+	protected void startObserve(final Request request) {
+
+		// We ignore blockwise request, except when this is an early negociation
+		// (num and M is set to 0)
+		if (!request.getOptions().hasBlock2() || request.getOptions().getBlock2().getNum() == 0
+				&& !request.getOptions().getBlock2().isM()) {
+			// add request to the store
+			final KeyToken idByToken = KeyToken.fromOutboundMessage(request);
+			LOG.log(Level.FINER, "registering observe request {0}", request);
+			observationStore.add(new Observation(request, null));
+			// remove it if the request is cancelled, rejected or timedout
+			request.addMessageObserver(new MessageObserverAdapter() {
+
+				@Override
+				public void onCancel() {
+					observationStore.remove(request.getToken());
+					exchangeStore.releaseToken(idByToken);
+				}
+
+				@Override
+				public void onReject() {
+					observationStore.remove(request.getToken());
+					exchangeStore.releaseToken(idByToken);
+				}
+
+				@Override
+				public void onTimeout() {
+					observationStore.remove(request.getToken());
+					exchangeStore.releaseToken(idByToken);
+				}
+			});
+		}
+	}
+
+	protected Exchange matchNotifyResponse(final ExchangeObserver exchangeObserver, final Response response,
+			final CorrelationContext responseContext) {
+
+		final Exchange.KeyToken idByToken = Exchange.KeyToken.fromInboundMessage(response);
+		Exchange exchange = null;
+
+		// we didn't find a message exchange for the token from the response
```

Comment:
This part of the comment should now go to the code fragment where this method is invoked from.